app/src/test/java/com/example/datastore/repository/DataStoreRepositoryTest.kt

```diff
@@ -1,88 +1,179 @@
 import org.hamcrest.MatcherAssert.assertThat
+import org.junit.Assert.fail
 import org.junit.Test
-import org.mockito.Mockito.`when`
-import org.mockito.Mockito.verify
+import org.mockito.Mockito.*
 
 private const val randomNumber = 4
 
 class DataStoreRepositoryTest {
 
-    private val dataStoreHelper = mock<DataStoreHelper>()
+    private val mockHelper = mock<DataStoreHelper>()
 
-    private val repository = DataStoreRepository(dataStoreHelper) { randomNumber }
+    private val repositoryWithMock = DataStoreRepository(mockHelper) { randomNumber }
+
+    private val fakeHelper = FakeDataStoreHelper()
+
+    private val repositoryWithFake = DataStoreRepository(fakeHelper) { randomNumber }
 
     @Test
     fun registerUser() = runBlocking {
         val user = StandardUser("test", "pass", randomNumber)
 
-        repository.register("test", "pass")
+        repositoryWithMock.register("test", "pass")
+
+        verify(mockHelper).addUser(user)
+    }
+
+    // USES MOCK
+
+    @Test
+    fun getUserDetails(): Unit = runBlocking {
+        val user = StandardUser("test", "test", 2)
+
+        repositoryWithMock.getUserDetails(user)
+
+        verify(mockHelper).getSingleUser(user)
+    }
+
+    @Test
+    fun generateMessage() = runBlocking {
+        val user = StandardUser("test", "test", 5)
+
+        repositoryWithMock.generateMessage(user)
 
-        verify(dataStoreHelper).addUser(user)
+        verify(mockHelper).updateUser(user.copy(message = randomNumber))
     }
 
     @Test
-    fun loginValidStandardUser() = runBlocking {
+    fun loginWaitsTillFirstCollectionOfUsers() = runBlockingWithCustomScope { _, customScope ->
         val user = StandardUser("test", "test", 2)
 
-        val flow = flow {
-            delay(20)
+        val delayedFlow = flow {
+            delay(100)
             emit(listOf(user))
         }
 
-        `when`(dataStoreHelper.users).thenReturn(flow)
-        val isValid = repository.login("test", "test")
+        `when`(mockHelper.users).thenReturn(delayedFlow)
+
+        repositoryWithMock.initUsers(customScope)
 
-        assertThat(isValid, `is`(user))
+        val result = repositoryWithMock.login("test", "test")
+
+        assertThat(result, `is`(user))
     }
 
     @Test
-    fun loginValidProtoBuffUser() = runBlocking {
-        val user = ProtoBuffUser("test", "test", 2, 0)
-        `when`(dataStoreHelper.users).thenReturn(flowOf(listOf(user)))
+    fun initUsersShouldLaunchOnlyOnce() = runBlockingWithCustomScope { _, customScope ->
+        val user = StandardUser("test", "pass", randomNumber)
+        `when`(mockHelper.users).thenReturn(flowOf(listOf(user)))
 
-        val isValid = repository.login("test", "test")
+        repositoryWithMock.initUsers(customScope)
+        repositoryWithMock.initUsers(customScope)
+        repositoryWithMock.initUsers(customScope)
 
-        assertThat(isValid, `is`(user))
+        // also i could test it via scope too by; but YAGNI
```

Comment:
i could also add a assert that compares launches coroutines in customScope; but YAGNI